ports/stm32/machine_i2s.c

```diff
@@ -584,13 +584,19 @@ STATIC bool i2s_init(machine_i2s_obj_t *self) {
 
     if (HAL_I2S_Init(&self->hi2s) == HAL_OK) {
         // Reset and initialize Tx and Rx DMA channels
+        uint32_t pm_size;
+        if (self->hi2s.Init.DataFormat == I2S_DATAFORMAT_16B) {
```

Comment:
Is this a bug fix or DMA optimization for 32-bit audio?

Reply:
This is a bug fix!  I'm surprised it works on other MCUs without this...

Reply:
Can you describe how the bug manifests?   I assume this bug was discovered during the H7 port.  Thanks.

Reply:
When using 32-bit RX (which is always the case, it's always 32-bit for incoming data from a mic) the values read by the DMA were always 0.  And the DMA filled its buffer twice as fast.  This makes sense because the DMA was doing only 16-bit reads from the I2S data register, and two such reads are needed per 32-bit sample.

Reply:
Also I'm pretty sure 32-bit TX was partially broken.  It did make sound on the speaker I used for testing, but it made "better" sound when I fixed this bug.

Reply:
Thanks.  That's clear now.  I'll take a closer look to see where I went wrong.  The I2S port to mimxrt is almost done.  I'll review the implementation to make sure the same bug doesn't show up on the mimxrt.

Reply:
After a more detailed look at the PR I do not see a bug in the original code:
1. The stm32 F4 and F7 documentation indicates that the I2S peripheral data width is a half-word. 
"DMA capability for transmission and reception (16-bit wide)" (32.3 of RM0385 Reference manual).   The I2S HAL docs discuss 16-bit transfers.
2. The Peripheral and Memory data width is specified as half-words in `dma.ca` in `dma_init_struct_i2s`.  
This structure data is used to initialize DMA in `dma_init_handle()` with `dma->Init = *dma_descr->init;`
3. Steps 1 and 2 would appear to correctly setup DMA for the F4 and F7 processor series.  

I am struggling to see a bug, at least for the F4 and F7 processors. The existing I2S implementation seems to follow
the manufacturers best practices.  

What information did you find to suggest a different approach to what is outlined above?  e.g.  using word size DMA.

Reply:
I think you are right, I think the original code is correct for F4 and F7.

It's on H7 that the I2S data register (which now has separate TXDR and RXDR registers) behaves differently.

![h7i2s](https://user-images.githubusercontent.com/6187689/153703275-15799566-4cb7-4c0a-9c62-81c2fe337038.png)

Reply:
That makes sense to me now. 

Here is an implementation idea:  An alternative to overriding the dma data width sizes in the dma init struct (eg  dma_init_struct_i2s in dma.c) with the new function dma_init_with_size() is to modify the init structs in dma.c.

For example:
`const dma_descr_t dma_I2S_2_TX = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s };`

would change to:
`const dma_descr_t dma_I2S_2_TX_16 = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s_16 };`
and
`const dma_descr_t dma_I2S_2_TX_32 = { DMA1_Stream3, DMA_CHANNEL_0, dma_id_3,   &dma_init_struct_i2s_32 };`

There would be separate init structs for 16-bit vs 32-bit sample sizes.  These init structs would have macros for F4/F7/H7/etc to conditionally compile for uniqueness in the processor families, as it does now.

in machine_i2s.c, the code that looks like this:
```
        // configure DMA streams
        if (self->mode == I2S_MODE_MASTER_RX) {
            self->dma_descr_rx = &dma_I2S_1_RX;
        } else {
            self->dma_descr_tx = &dma_I2S_1_TX;
        }
```
would change to:

```
        // configure DMA streams
        if (self->mode == I2S_MODE_MASTER_TX && get_dma_bits(self->mode, self->bits) ==16  ) {
            self->dma_descr_rx = &dma_I2S_1_TX_16;
        } else if  (self->mode == I2S_MODE_MASTER_TX && get_dma_bits(self->mode, self->bits) ==32  ){
            self->dma_descr_tx = &dma_I2S_1_TX_32;
        } else if RX {
            self->dma_descr_rx = &dma_I2S_1_RX_32;    --- always 32 bits
        }
```
If this works, it has the advantage of keeping all the dma data width information in one place, dma.c -- which might aid in self-documenting the code.

Reply:
Another consideration for the H7 are the transformations that happen in machine_i2s.c code:
1.  transform for 32-bit sample TX:  https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/machine_i2s.c#L246
2. "cherry-picking" transform for RX: https://github.com/micropython/micropython/blob/2ea21abae0358a46c0144f014f2e5891afe03dfd/ports/stm32/machine_i2s.c#L308

You mentioned that master TX has been tested, so that should eliminate item 1 as a concern.  Hopefully the transform in item 2 will work as-is for the H7.

Reply:
Yes I like your idea about having multiple DMA descriptor structs, for 16 and 32 bit transfers.   I'll try that out.

And will test again all combinations of TX, RX and bit size on H7.

Reply:
I changed the DMA config so it uses separate 16 and 32 bit descriptor structs.

And I tested the byte transformations, they were wrong on the H7, and are now fixed (at least for 32-bit stereo).